Stratis.Bitcoin.Features.Consensus/ConsensusLoop.cs

```diff
@@ -78,41 +90,56 @@ public BlockResult ExecuteNextBlock(CancellationToken cancellationToken)
                     {
                         result.Block = this.Puller.NextBlock(cancellationToken);
                         if (result.Block != null)
+                        {
+                            this.logger.LogTrace("Block received from puller.");
                             break;
-                        else
+                        }
+
+                        this.logger.LogTrace("No block received from puller due to reorganization, rewinding.");
+                        while (true)
                         {
-                            while (true)
+                            uint256 hash = this.UTXOSet.Rewind().GetAwaiter().GetResult();
+                            ChainedBlock rewinded = this.Chain.GetBlock(hash);
+                            if (rewinded == null)
                             {
-                                var hash = this.UTXOSet.Rewind().GetAwaiter().GetResult();
-                                var rewinded = this.Chain.GetBlock(hash);
-                                if (rewinded == null)
-                                    continue;
-                                this.Tip = rewinded;
-                                this.Puller.SetLocation(rewinded);
-                                break;
+                                this.logger.LogTrace("Rewound to '{0}', which is still not a part of the current best chain, rewinding further.", hash);
+                                continue;
                             }
+
+                            this.logger.LogTrace("Rewound to '{0}'.", hash);
+                            this.Tip = rewinded;
+                            this.Puller.SetLocation(rewinded);
+                            break;
                         }
                     }
                 }
 
                 this.AcceptBlock(new ContextInformation(result, this.Validator.ConsensusParams));
             }
-            catch(ConsensusErrorException ex)
+            catch (ConsensusErrorException ex)
             {
                 result.Error = ex.ConsensusError;
             }
 
+            this.logger.LogTrace("(-):*.{0}='{1}/{2}',*.{3}='{4}'", nameof(result.ChainedBlock), result.ChainedBlock?.HashBlock, result.ChainedBlock?.Height, nameof(result.Error), result.Error?.Message);
             return result;
         }
 
         public void AcceptBlock(ContextInformation context)
         {
+            this.logger.LogTrace("()");
+
             using (new StopwatchDisposable(o => this.Validator.PerformanceCounter.AddBlockProcessingTime(o)))
             {
                 // Check that the current block has not been reorged.
                 // Catching a reorg at this point will not require a rewind.
                 if (context.BlockResult.Block.Header.HashPrevBlock != this.Tip.HashBlock)
+                {
+                    this.logger.LogTrace("Reorganization detected.");
```

Comment:
I am not sure this is a reorg, its rather a bug the wrong block was sent in to the consensus engine.
Its a requirement that the next block must always be pushed.

Reply:
Sometimes I tend to follow the inline comment when I insert logs to code I am not that familiar with. So if I read that we are catching a reorg here, I just write reorg to the log. 

However, when this actually happened in real life to my node, it was during reorg, so I believe it is possibly correct here. If there is another possibility in case of bug, that is not very interesting as the log should reflect the correct and expected behavior.